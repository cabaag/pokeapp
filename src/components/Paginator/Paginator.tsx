import { Button, Icon, Text } from 'native-base';
import React, { memo, useCallback, useEffect, useState } from 'react';
import { useTranslation } from 'react-i18next';
import { StyleSheet, View } from 'react-native';

const styles = StyleSheet.create({
  paginator: {
    flexDirection: "row",
    flex: 1,
    left: 0,
    right: 0,
    justifyContent: 'space-between',
    padding: 8
  }
})

export interface PaginatorProps {
  count: number;
  loading: boolean;
  onChangePage: (page: number) => void;
  interval?: number;
}

const Paginator: React.FC<PaginatorProps> = ({
  count,
  loading,
  onChangePage,
  interval = 5,
}: PaginatorProps) => {

  const [currentPage, setCurrentPage] = useState<number>(0);
  const [lastPage, setLastPage] = useState(0);
  const [offset] = useState(Math.floor(interval / 2));

  const handleChangePage = useCallback((page: number) => {
    setCurrentPage(page);
    onChangePage(page);
  }, [onChangePage]);

  useEffect(() => {
    setLastPage(Math.floor(count / 50))
  }, [count])

  const renderButtons = useCallback(() => {
    let minPage: number = Math.min(...[
      currentPage - offset, Math.abs(lastPage - interval)
    ])
    minPage = minPage < 0 ? 0 : minPage;
    let maxPage: number = Math.min(...[currentPage + interval - offset, lastPage])
    maxPage = maxPage < 5 ? 5 : maxPage;
    const buttons = [];

    // eslint-disable-next-line no-plusplus
    for (let i = minPage; i < maxPage; i++) {
      buttons.push(
        <Button bordered={currentPage !== i} disabled={currentPage === i || loading} key={i} onPress={() => handleChangePage(i)} test-id={`page-button-${i}`}>
          <Text>
            {i + 1}
          </Text>
        </Button>
      )
    }
    return buttons;
  }, [lastPage, currentPage, loading])

  const prevDisabled = currentPage === 0 || loading;
  const nextDisabled = currentPage === lastPage - 1 || loading;
  return (
    <View style={styles.paginator}>
      <Button bordered disabled={prevDisabled} onPress={() => handleChangePage(0)} test-id="first">
        <Icon name="previous" type="Foundation" />
      </Button>
      { renderButtons()}
      <Button bordered disabled={nextDisabled} onPress={() => handleChangePage(lastPage - 1)} test-id="last">
        <Icon name="next" type="Foundation" />
      </Button>
    </View>
  )
}

export default memo(Paginator);