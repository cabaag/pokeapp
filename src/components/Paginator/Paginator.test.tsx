import { shallow, ShallowWrapper } from 'enzyme';
import React from 'react';
import Paginator, { PaginatorProps } from './Paginator';

const onChangePage = jest.fn();

jest.mock('react-i18next', () => ({
  useTranslation: () => ({ t: (key: string) => key })
}));

const defaultProps: PaginatorProps = {
  count: 100,
  loading: false,
  onChangePage
}

function setup(customProps?: Partial<PaginatorProps>): ShallowWrapper {
  return shallow(
    <Paginator {...defaultProps} {...customProps} />
  )
}

describe('<Paginator />', () => {
  let wrapper: ShallowWrapper;


  beforeEach(() => {
    wrapper = setup();
  })

  it('renders correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('renders buttons', () => {
    const buttons = wrapper.find("[test-id='page-button']")
    expect(buttons).toHaveLength(5);
  })

  it('disable buttons', () => {
    onChangePage.mockReset();
    wrapper = setup({
      loading: true,
    });
    const nextButton = wrapper.find("[test-id='next']")
    nextButton.simulate('click')
    expect(onChangePage).toHaveBeenCalledTimes(0);
  })

});