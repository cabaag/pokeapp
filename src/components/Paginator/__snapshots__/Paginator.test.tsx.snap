// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Paginator /> renders correctly 1`] = `
<View
  style={
    Object {
      "flex": 1,
      "flexDirection": "row",
      "justifyContent": "space-between",
      "left": 0,
      "padding": 8,
      "right": 0,
    }
  }
>
  <Styled(Button)
    disabled={true}
    onPress={[Function]}
    test-id="prev"
  >
    <Styled(Text)>
      previous
    </Styled(Text)>
  </Styled(Button)>
  <Styled(Button)
    disabled={false}
    key="-3"
    onPress={[Function]}
    test-id="page-button"
  >
    <Styled(Text)>
      -2
    </Styled(Text)>
  </Styled(Button)>
  <Styled(Button)
    disabled={false}
    key="-2"
    onPress={[Function]}
    test-id="page-button"
  >
    <Styled(Text)>
      -1
    </Styled(Text)>
  </Styled(Button)>
  <Styled(Button)
    disabled={false}
    key="-1"
    onPress={[Function]}
    test-id="page-button"
  >
    <Styled(Text)>
      0
    </Styled(Text)>
  </Styled(Button)>
  <Styled(Button)
    disabled={true}
    key="0"
    onPress={[Function]}
    test-id="page-button"
  >
    <Styled(Text)>
      1
    </Styled(Text)>
  </Styled(Button)>
  <Styled(Button)
    disabled={false}
    key="1"
    onPress={[Function]}
    test-id="page-button"
  >
    <Styled(Text)>
      2
    </Styled(Text)>
  </Styled(Button)>
  <Styled(Button)
    disabled={false}
    onPress={[Function]}
    test-id="next"
  >
    <Styled(Text)>
      next
    </Styled(Text)>
  </Styled(Button)>
</View>
`;
